### ISSUE TITLE
discrepancy with running mhcflurry-predict on not supported alleles

### ISSUE THREAD

**Issue by liviatran** [openvax/mhcflurry#210]
I ran `mhcflurry-predict --list-supported-alleles > 'supported_alleles.txt'` to get a list of of alleles supported for mhcflurry. 

I noted HLA-A*02:172, HLA-C*02:16, and HLA-C12:139 were not in that list of supported alleles. The aforementioned alleles are all Well-Documented alleles (https://www.ihiw18.org/component-immunogenetics/download-common-and-well-documented-alleles-3-0).

I was able to successfully run mhcflurry-predict on HLA-A*02:172 and HLA-C*02:16, despite them not being on the list of supported alleles. HLA-C12:139 caused a ValueError.

Are the results for the unsupported alleles (HLA-A*02:172, HLA-C*02:16) reliable?

**Comment by timodonnell** [comment]
Looks like the first two alleles you mentioned are getting canonicalized by mhcgnomes to other allele names, whereas the third is genuinely unsupported:

<img width="1188" alt="image" src="https://user-images.githubusercontent.com/467896/206871758-62bc8699-1b98-48d3-b4e9-a91fa1e1c00b.png">

We should take a look at the sequences in IMGT for the first two alleles and what they are getting mapped to (HLA-A02:172 -> HLA-A*02:17, HLA-C02:16 -> HLA-C*02:137) to understand if this canonicalization is reasonable or a bug. I will add this to my todo list but if you have a chance to do it first please let us know what you see. @iskandr who wrote mhcgnomes may also have thoughts on this.

**Comment by liviatran** [comment]
I'm not sure why A*02:172 is getting mapped to A*02:17, as they are different alleles with different protein sequences.

However, I did note that C*02:16:01 was renamed to C*02:137 from IMGT's Deleted_alleles.txt

<img width="441" alt="Screen Shot 2022-12-15 at 9 45 55 PM" src="https://user-images.githubusercontent.com/32661491/208031495-507ecd0d-48d8-467f-b003-b3adb9d06814.png">

As for C*12:139, this allele was changed from C*12:139 in IMGT v3.38.0 to C*12:139Q in v 3.39.0.

**Comment by liviatran** [comment]
I'm not sure why A*02:172 is getting mapped to A*02:17, as they are different alleles with different protein sequences.

I did note that C*02:16:01 was renamed to C*02:137 from IMGT's Deleted_alleles.txt. However, C*02:16 shouldn't be mapped to C*012:137.

<img width="441" alt="Screen Shot 2022-12-15 at 9 45 55 PM" src="https://user-images.githubusercontent.com/32661491/208031495-507ecd0d-48d8-467f-b003-b3adb9d06814.png">

As for C*12:139, this allele was changed from C*12:139 in IMGT v3.38.0 to C*12:139Q in v 3.39.0.

**Comment by timodonnell** [comment]
@iskandr do you think the following is a mhcgnomes bug:

```
> mhcgnomes.parse("HLA-A*02:172", use_allele_aliases=True).restrict_allele_fields(2).to_string()
HLA-A*02:17
```

@liviatran I am seeing that C02:16:01 is getting canonicalized to HLA-C02:137, which from your comment I think is the correct canonicalization, right? (I.e. I am not seeing it getting canonicalized to C012:137.)

**Comment by iskandr** [comment]
This does seem like a bug. 

```
In [1]: mhcgnomes.parse("HLA-A*02:172", use_allele_aliases=True)
Out[2]: Allele(gene=Gene(species=Species(name='Homo sapiens', mhc_prefix='HLA'), name='A', mutations=()), allele_fields=('02', '17', '02', '01'), annotations=(), mutations=())
```
Checking the IMGT/HLA allele history entry, I see:

```
HLA03784,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*02:172,A*9272,A*9272,A*9272,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA
```

I'm trying to figure out how this "normalization" happens and my guess is that it's via this entry:

```
HLA00023,A*02:17:02:01,A*02:17:02:01,A*02:17:02:01,A*02:17:02:01,A*02:17:02:01,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*02:17:02,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*021702,A*02172,A*02172,A*02172,A*02172,A*02172,A*02172,A*02172,A*02172,A*02172,A*02172,A*02172
```

...which links "A*02172" with "A*02:17:02:01". 

I'm going to figure out where in the logic "A*02:172" gets checked against "A*02172" and make it more cautious.

**Comment by liviatran** [comment]
@timodonnell I don't believe C*02:16 should be canonicalized to C*02:137. C*02:16:01 was renamed to C*02:137 since it encodes an entirely different protein than other C*02:16 alleles.

**Comment by liviatran** [comment]
@iskandr Perhaps mhcgenomes.parse should only look at version 3 (colon delimited) HLA allele names to avoid these name collisions. 

For the non-colon delimited (versions 1 and 2) allele names, the names have to be evaluated in pairs of numbers. There would never be a version 1 or 2 allele name with three digits in the first field.

**Comment by iskandr** [comment]
Resolved in current releases and verified with v2.3.0rc14 code plus the public predictor. HLA-C*02:16 and HLA-C*12:139 resolve to supported keys and score successfully. mhcgnomes identifies HLA-A*02:172 as a historical alias of HLA-A*02:17:02:01; MHCflurry intentionally uses the protein-resolution key HLA-A*02:17, which is supported and scores successfully.
